### PR TITLE
Laboratory work #3, Lopatko Yulia - 21FPL1

### PR DESCRIPTION
**Pull request by lazykosti**

### CODE REVIEW

**Comment by lazykosti** [comment]
problem with solving unittests' failing (it seems like those tests that i didn't write are failing, for example, steps 8.2-8.4)

**Comment by artyomtugaryov** [comment]
Write to our chat about problem with tests

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -120,7 +127,9 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        for index, element in enumerate(tokens):
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:130]: read about `start` argument

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -173,7 +193,16 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
     tokens must not be empty, window lengths must be integer, window lengths cannot be less than 2.
     """
-    pass
+    if not tokens or not isinstance(window_length, int) or window_length < 2:
+        return None
+    pairs = []
+    for index in range(len(tokens) - window_length):
+        for i in range(window_length):
+            for j in range(i, window_length):
+                for pair in [(tokens[index + i], tokens[index + j])]:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:202]: looks like this for is unnecessary - iteration is always in one element

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -233,7 +264,21 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        vertexes = [vertex1, vertex2]
+        for vertex in vertexes:
+            if vertex not in self._matrix[0]:
+                self._matrix[0].append(vertex)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:272]: again

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -136,7 +145,11 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+        self._learn_indices(tokens)
+        encoded_tokens = tuple(self._word2id.values())
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:151]: I think, it is a wrong implementation.

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -152,7 +165,14 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        tokens = []
+        for token in encoded_tokens:
+            if token not in self._id2word:
+                return None
+            for number in self._id2word:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:172]: why don't you just do `tokens.append(self._id2word[number])` ? Why do you need this loop?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (artyomtugaryov) [reply lab_3_keywords_textrank/main.py:172]: ?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**